### PR TITLE
Hybrid Smart Graphs: Adds support for clearing Providers

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

Enumerators are responsible for clearing Providers as well.
Both available Providers (OneSided and TwoSided) got new methods to verify integrity.
ProviderTracer / MockGraphProvider got clear() method due to interface addition.

- [x] :hammer: Refactoring/simplification

#### Backports:

- [x] No backports required

#### Related Information

- [x] Design document: https://arangodb.atlassian.net/secure/RapidBoard.jspa?rapidView=55&modal=detail&selectedIssue=GORDO-1230

### Testing & Verification

- [x] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **ASSERTS**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by hkernbach** [comment]
ASAN run started:
- https://jenkins.arangodb.biz:3456/job/arangodb-ANY-linux-asan/756/

**Review by mchacki — CHANGES_REQUESTED** [review]
Technical production code looks good to me, this is a good change making everything more secure and avoid Undefined behavior.

However please move the clearProvider methods into private state, as they are unsafe to be called from outside, and never intended to be.

File: arangod/Aql/ExecutionEngine.cpp [hunk 1]
```diff
@@ -644,6 +644,7 @@ void ExecutionEngine::instantiateFromPlan(Query& query,
   ExecutionEngine* engine = nullptr;
 #ifdef USE_ENTERPRISE
   bool const pushToSingleServer = plan.hasAppliedRule(
+      // Feature HybridSmartGraphs: Check here.
```
Comment (hkernbach) [inline arangod/Aql/ExecutionEngine.cpp:647]:
```suggestion
```

File: arangod/Aql/QuerySnippet.cpp [hunk 2]
```diff
@@ -686,12 +686,19 @@ auto QuerySnippet::prepareFirstBranch(
         }
       }
 
+      // TODO: We need to exclude DBServers which only do have Satellite
+      // collections only This server is not allowed to receive a setup call
+
+      // 1.) Satellites: 1x shard exists (681) -> not landing in myExpFinal
+      // => 2x shards vertex collections -> DB1 v: [s1, s2] , DB2 []
+      // 2.)
+
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:695]:
```suggestion
```
That is already handled.

File: arangod/Aql/QuerySnippet.cpp [hunk 3]
```diff
@@ -686,12 +686,19 @@ auto QuerySnippet::prepareFirstBranch(
...
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
       // additional verification checks for Disjoint SmartGraphs
       if (localGraphNode->isDisjoint()) {
         if (!myExpFinal.empty()) {
           size_t numberOfShards = myExpFinal.begin()->second.size();
-          // We need one expansion for every non-satellite collection in the Graph
+          // We need one expansion for every non-satellite collection in the Graph (-1 per satellite)
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:701]:
```suggestion
          // We need one expansion for every non-satellite collection in the Graph
```
Non satellite in the comment already.

File: arangod/Aql/ShardLocking.cpp [hunk 4]
```diff
@@ -108,7 +108,7 @@ void ShardLocking::addNode(ExecutionNode const* baseNode, size_t snippetId,
       }
       auto const graphIsUsedAsSatellite = graphNode->isUsedAsSatellite();
       auto const isUsedAsSatellite = [&](auto const& col) {
-        return graphIsUsedAsSatellite || (col->isSatellite() && (pushToSingleServer || graphNode->isSmart()));
+        return graphIsUsedAsSatellite || (col->isSatellite() && (pushToSingleServer || graphNode->isSmart())) || (col->isSatellite() && graphNode->isSmart());
```
Comment (mchacki) [inline arangod/Aql/ShardLocking.cpp:111]:
```suggestion
        return graphIsUsedAsSatellite || (col->isSatellite() && (pushToSingleServer || graphNode->isSmart())));
```
Covered by the second OR case

File: arangod/ClusterEngine/ClusterIndexFactory.cpp [hunk 5]
```diff
@@ -78,15 +77,14 @@ struct DefaultIndexFactory : public IndexTypeFactory {
...
-                                          definition);
+                                          Index::type(_type), definition);
   }
 
   virtual Result normalize(
       velocypack::Builder& normalized,
       velocypack::Slice definition,
       bool isCreation,
-      TRI_vocbase_t const& vocbase) const override {
+      TRI_vocbase_t const& vocbase             ) const override {
```
Comment (mchacki) [inline arangod/ClusterEngine/ClusterIndexFactory.cpp:87]:
```suggestion
      TRI_vocbase_t const& vocbase) const override {
```

File: arangod/ClusterEngine/ClusterIndexFactory.cpp [hunk 6]
```diff
@@ -98,7 +96,7 @@ struct DefaultIndexFactory : public IndexTypeFactory {
     }
 
     return engine->indexFactory().factory(_type).normalize(
-      normalized, definition, isCreation, vocbase);
+        normalized, definition, isCreation, vocbase          );
```
Comment (mchacki) [inline arangod/ClusterEngine/ClusterIndexFactory.cpp:99]:
```suggestion
        normalized, definition, isCreation, vocbase);
```

File: arangod/ClusterEngine/ClusterIndexFactory.cpp [hunk 7]
```diff
@@ -213,7 +211,7 @@ Result ClusterIndexFactory::enhanceIndexDefinition(  // normalize definition
   }
 
   return ae->indexFactory().enhanceIndexDefinition(
-    definition, normalized, isCreation, vocbase);
+      definition, normalized, isCreation, vocbase    );
```
Comment (mchacki) [inline arangod/ClusterEngine/ClusterIndexFactory.cpp:214]:
```suggestion
      definition, normalized, isCreation, vocbase);
```

File: arangod/Graph/BreadthFirstEnumerator.cpp [hunk 8]
```diff
@@ -166,6 +166,12 @@ bool BreadthFirstEnumerator::next() {
           return;
         }
 
+#ifdef USE_ENTERPRISE
+        if (!validDisjointPath(nextIdx, vId)) {
+          return;
+        }
+#endif
+
```
Comment (mchacki) [inline arangod/Graph/BreadthFirstEnumerator.cpp:174]:
```suggestion
```
That code is already there above.

File: arangod/Graph/PathEnumerator.cpp [hunk 9]
```diff
@@ -288,6 +288,12 @@ bool DepthFirstEnumerator::next() {
           return;
         }
 
+#ifdef USE_ENTERPRISE
+        if (!validDisjointPath()) {
+          return;
+        }
+#endif
+
```
Comment (mchacki) [inline arangod/Graph/PathEnumerator.cpp:296]:
```suggestion
```
Already in the lines above

File: arangod/Graph/Providers/ClusterProvider.h [hunk 10]
```diff
@@ -140,6 +140,7 @@ class ClusterProvider {
     Vertex _vertex;
     Edge _edge;
     bool _fetched;
+    size_t _localSchreierIndex = std::numeric_limits<size_t>::max();  // to be removed later
```
Comment (mchacki) [inline arangod/Graph/Providers/ClusterProvider.h:143]: Is this needed?
I do not see related .cpp code.
maybe this was made obsolete in the original.

**Review by mchacki — APPROVED** [review]
Technically good, there are some duplicates that can be removed via my inline suggestions.
Mostly comments / whitespace.

**Comment by mchacki** [comment]
The Red part in Jenkins was arm-build that could not connect to GitHub.
We have seen it green before with non relevant changes.

**Pull request merged**